WalletWasabi/Crypto/ZeroKnowledge/ProofSystem.cs

```diff
@@ -65,5 +67,123 @@ public static Knowledge BalanceProof(Scalar zSum, Scalar rDeltaSum)
+			var bitCommitments = bits.Zip(randomness, (b, r) => b * Generators.Gg + r * Generators.Gh);
+
+			var columns = width * 3 + 1; // three witness terms per bit and one for the commitment
+			int bitColumn(int i) => 3 * i + 1;
+			int rndColumn(int i) => bitColumn(i) + 1;
+			int productColumn(int i) => bitColumn(i) + 2;
+
+			// Construct witness vector. First term is r from Ma = a*Gg + r*Gh. This
+			// is followed by 3 witness terms per bit commitment, the bit b_i, the
+			// randomness in its bit commitment r_i, and their product rb_i (0 or r).
+			var witness = new Scalar[columns];
+			witness[0] = r;
+			foreach ((Scalar b_i, Scalar r_i, int i) in bits.Zip(randomness, Enumerable.Range(0, width), (x, y, z) => (x, y, z)))
+			{
+				witness[bitColumn(i)] = b_i;
+				witness[rndColumn(i)] = r_i;
+				witness[productColumn(i)] = r_i * b_i;
+			}
+
+			return (new Knowledge(RangeProof(ma, bitCommitments), new ScalarVector(witness)), bitCommitments);
+		}
+
+		// overload for bootstrap credential request proofs.
+		// this is just a range proof with width=0
+		// equivalent to new Statement(ma, Generators.Gh)
+		public static Statement ZeroProof(GroupElement ma)
+			=> RangeProof(ma, new GroupElement[0]);
+
+		public static Statement RangeProof(GroupElement ma, IEnumerable<GroupElement> bitCommitments)
+		{
+			var width = bitCommitments.Count();
+			Guard.True(nameof(width), width >= 0); // a width of 0 means no bits are set (commitment to 0)
+			Guard.True(nameof(width), width <= 255); // 256 is tautological for the order of secp256k1
+
+			var rows = width * 2 + 1; // two equations per bit, and one for the sum
+			var columns = width * 3 + 1 + 1; // three witness components per bit and one for the Ma randomness, plus one for the public inputs
+
+			// uninitialized values are implicitly O
+			var equations = new GroupElement[rows, columns];
+
+			// Proof of [ ( \sum 2^i * B_i ) - Ma = (\sum2^i r_i)*Gh - r*Gh ]
+			// This means that the bit commitments, if they are bits (proven in
+			// subsequent equations), are a decomposition of the amount committed in
+			// Ma, proven by showing that the public input is a commitment to 0 (only
+			// Gh term required to represent it). The per-bit witness terms of this
+			// equation are added in the loop below.
+			var bitsTotal = bitCommitments.Select((B, i) => Scalar.Zero.CAddBit((uint)i, 1) * B).Sum();
+			equations[0, 0] = ma - bitsTotal;
+			equations[0, 1] = Generators.Gh; // first witness term is r in Ma = a*Gg + r*Gh
+
+			// Some helper functions to calculate indices of witness terms.
+			// The witness structure is basically: r, zip3(b_i, r_i, rb_i)
+			// So the terms in each equation are the public input (can be thought of
+			// as a generator for a -1 term ) and the remaining are generators to
+			// be used with 1+3n terms:
+			//   ( r, b_0, r_0, rb_0, b_1, r_1, rb_1, ..., b_n, r_n, rb_n)
+			int bitColumn(int i) => 3 * i + 2; // column for b_i witness term
```

Comment:
@kiminuo what do you think about this local functions? Do you have any idea about how to improve this? My attempt is here: https://github.com/zkSNACKs/WalletWasabi/pull/4431/file

Reply:
The problem in a nutshell: we have a bunch of variables (columns, terms in the witness) that plug into a system of equations (the rows). there's 2 equations and 3 columns per bit in the range proof, so each bit needs 6 cells on its own, together or apart, doesn't matter, and then there's one more equation which uses terms from all of the bits (but not all terms).

A third approach which seems overcomplicated to me but might help at this level of abstraction (by adding like 47 more), we could express this relationship more symbolically: In `Statement`'s constructor use bindings generated by low level helpers into an `IEnumerable<(ScalarIdentifier id, GroupElement generator)> where `ScalarIdentifier` is just some unique identifier (e.g. an incrementing number). Then add a variable unification operation on `ScalarIdentifiers` that declares two terms equal which could then be used to compose the symbolic equations constructed by these objects. This structure would then need to be flattened into a tabular form, as `Equation` objects or a `GroupElement[,]`, or become the direct underlying representation of `Statement`/`Equation`.

A fourth approach which seems kind of hacky to me is to use a dictionary/map/hashmap/whatever they're called in C# and just generate variable names for things. Gross.

I can't think of another idea, but I these ones are bad enough ;-)

I also tried tried to orient things differently as `( r, b_0, b_1, .. b_n, r_0, r_1, ... r_n, rb_0, .. rb_n )` but it was no simpler, in both cases moving the index calculations directly into the `[]` operation is too opaque in my opinion, this code already smells enough like C.

Reply:
@lontivero The need for local functions seems like a hint that you need a new class where you can convert those local functions to normal methods (or static ones). So I would simply add new `RangeProof` class.

Reply:
Also if those local functions cannot be re-used I would just convert them to simple variables. I find it ok even when names are not that great but when you add some comment specifying what it is, it should be good. 

It's math, there are some formulae that are complex and you can't really simplify them.

Reply:
regarding creating a new class - i don't think that would really change very much here, since that class would have to implement this same logic of creating a table with these interleaved rows and columns then it just moves the ugliness that around to there, but i guess yeah i haven't been able to think of a simpler way for more than a week now so i'm ready to call it as good as it gets for now

fwiw they are used multiple times in the loop with different numbers for the bits